cpp/ql/src/Security/CWE/CWE-311/CleartextBufferWrite.ql

```diff
@@ -12,31 +12,39 @@
  */
 
 import cpp
-import semmle.code.cpp.security.BufferWrite
+import semmle.code.cpp.security.BufferWrite as BufferWrite
 import semmle.code.cpp.security.SensitiveExprs
 import semmle.code.cpp.security.Security
-import semmle.code.cpp.dataflow.TaintTracking
+import semmle.code.cpp.ir.dataflow.TaintTracking
 import DataFlow::PathGraph
 
+Expr exprForNode(DataFlow::Node n) {
+  n = DataFlow::exprNode(result)
+  or
+  // (similar to DefaultTaintTracking's `getNodeForExpr`)
+  n = DataFlow::definitionByReferenceNodeFromArgument(result) and
+  not argv(result.(VariableAccess).getTarget())
+}
+
 /**
  * Taint flow from user input to a buffer write.
  */
 class ToBufferConfiguration extends TaintTracking::Configuration {
   ToBufferConfiguration() { this = "ToBufferConfiguration" }
 
-  override predicate isSource(DataFlow::Node source) { isUserInput(source.asExpr(), _) }
+  override predicate isSource(DataFlow::Node source) { isUserInput(exprForNode(source), _) }
```

Comment:
<s>Do we get the same results if we use `source instanceof FlowSource`? Then we don't need this ugly `exprForNode` (as `FlowSource.getSourceType` can be used instead of `isUserInput`  to get the `taintCause`).</s> Oh, you did exactly that in the last commit. Excellent!